src/cargo/ops/cargo_output_metadata.rs

```diff
@@ -0,0 +1,185 @@
+}
+
+/// Loads the manifest, resolves the dependencies of the project to the concrete
+/// used versions - considering overrides - and writes all dependencies in a TOML
+/// format to stdout or the specified file.
+///
+/// The TOML format is e.g.:
+/// ```toml
+/// root = "libA"
+///
+/// [packages.libA]
+/// dependencies = ["libB"]
+/// path = "/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/libA-0.1"
+/// version = "0.1"
+///
+/// [packages.libB]
+/// dependencies = []
+/// path = "/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/libB-0.4"
+/// version = "0.4"
+///
+/// [packages.libB.features]
+/// featureA = ["featureB"]
+/// featureB = []
+/// ```
+pub fn output_metadata(opt: OutputMetadataOptions, config: &Config) -> CargoResult<()> {
+    let (resolved_deps, packages) =
+        try!(resolve_dependencies(opt.manifest_path, opt.features, config));
+
+	#[derive(RustcEncodable)]
+    struct PackageInfo<'a> {
+        version: String,
+        path: Cow<'a, str>,
+        dependencies: Vec<String>,
+        features: Option<&'a HashMap<String, Vec<String>>>,
+        targets: &'a[Target]
+    };
+
+	#[derive(RustcEncodable)]
+	struct ExportInfo<'a> {
+        root: String,
+		packages: HashMap<String, PackageInfo<'a>>
+	}
+
+    let mut output = ExportInfo {
+        root: resolved_deps.root().name().to_string(),
+        packages: HashMap::new()
+	};
+
+    for package in packages.iter() {
+        let mut package_info = PackageInfo {
+            version: format!("{}", package.version()),
+            path: (*package.root()).to_string_lossy(),
+            dependencies: Vec::new(),
+            features: {
+                let features = package.manifest().summary().features();
+                if features.is_empty() {
+                    None
+                } else {
+                    Some(features)
+                }
+            },
+            targets: package.manifest().targets()
+        };
+
+        if let Some(dep_deps) = resolved_deps.deps(package.package_id()) {
+            for dep_dep in dep_deps {
+                package_info.dependencies.push(dep_dep.name().to_string());
+            }
+        }
+
+        output.packages.insert(package.name().to_string(), package_info);
+    }
+
+    let serialized_str = match opt.output_format {
+		OutputFormat::Toml => toml::encode_str(&output),
+		OutputFormat::Json => try!(json::encode(&output)),
+	};
+
+    match opt.output_to {
+        OutputTo::StdOut         => try!(config.shell().say(serialized_str, BLACK)),
+        OutputTo::Path(ref path) => {
+            let mut file = try!(fs::File::create(path));
+            try!(file.write_all(serialized_str.as_bytes()));
+        }
+    }
+
+    Ok(())
+}
+
+/// Loads the manifest and resolves the dependencies of the project to the
+/// concrete used versions. Afterwards available overrides of dependencies are applied.
+fn resolve_dependencies(manifest: &Path, features: Vec<String>, config: &Config)
```

Comment:
This function looks like it's got a good bit of duplication from pieces elsewhere throughout the codebase, can this duplication be removed?